source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -160,18 +160,32 @@ Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderM
 
 Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
   ENVOY_STREAM_LOG(trace,
-                   "ext_authz filter has {} response header(s) to add and {} response header(s) to "
-                   "set to the encoded response:",
-                   *encoder_callbacks_, response_headers_to_add_.size());
-  if (!response_headers_to_add_.empty()) {
+                   "ext_authz filter has {} response header(s) to append, {} response header(s) "
+                   "to add, {} response header(s) to set to the encoded response:",
+                   *encoder_callbacks_, response_headers_to_append_.size(),
+                   response_headers_to_add_if_absent_.size(), response_headers_to_set_.size());
+  // Headers to be added only if they are absent.
+  if (!response_headers_to_add_if_absent_.empty()) {
+    ENVOY_STREAM_LOG(trace, "ext_authz filter add header(s) if absent to the encoded response:",
+                     *encoder_callbacks_);
+    for (const auto& header : response_headers_to_add_if_absent_) {
+      if (headers.get(header.first).empty()) {
+        ENVOY_STREAM_LOG(trace, "'{}':'{}'", *encoder_callbacks_, header.first.get(),
+                         header.second);
+        headers.setCopy(header.first, header.second);
```

Comment:
Use `addCopy` here. `setCopy` will make an additional unnecessary `remove` which need to searching whole header map.

Reply:
Changed. Thanks!